paddle/fluid/operators/center_loss_op.cc

```diff
@@ -0,0 +1,152 @@
+/* Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/fluid/operators/center_loss_op.h"
+#include <memory>
+#include <string>
+
+namespace paddle {
+namespace operators {
+class CenterLossOp : public framework::OperatorWithKernel {
+ public:
+  CenterLossOp(const std::string &type,
+               const framework::VariableNameMap &inputs,
+               const framework::VariableNameMap &outputs,
+               const framework::AttributeMap &attrs)
+      : OperatorWithKernel(type, inputs, outputs, attrs) {}
+
+  void InferShape(framework::InferShapeContext *ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("X"),
+                   "Input(X) of CenterLoss should not be null.");
+    auto x_dims = ctx->GetInputDim("X");
+
+    PADDLE_ENFORCE(ctx->HasInput("CenterUpdateRate"),
+                   "Input(CenterUpdateRate) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(ctx->HasInput("Label"),
+                   "Input(Label) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(ctx->HasInput("Centers"),
+                   "Input(Centers) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(ctx->HasOutput("CentersDiff"),
+                   "Output(CentersDiff) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(ctx->HasOutput("Loss"),
+                   "Output(Loss) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(
+        ctx->HasOutput("CentersOut"),
+        "Output(CentersOut) of CenterLoss shared data with Centers.");
+
+    ctx->SetOutputDim("CentersDiff", {x_dims[0], product(x_dims) / x_dims[0]});
```

Comment:
Is the output dimension is just the x_dims?

Reply:
done